File: src/1617_Count_Subtrees_With_Max_Distance_Between_Cities.py
# 1617_Count_Subtrees_With_Max_Distance_Between_Cities
# https://leetcode.cn/problems/count-subtrees-with-max-distance-between-cities/solution/tu-jie-on3-mei-ju-zhi-jing-duan-dian-che-am2n/

# 方法一：枚举子集+树的直径
# 前置知识：子集型回溯

# 前置知识：二叉树的直径
# 见 543. 二叉树的直径。
# 推荐从这题开始学习树形 DP。
# 简单来说就是枚举从往左儿子走的最长链和往右儿子走的最长链，这两条链可能会组成直径。枚举所有点作为 x 就能找到答案。
# 每个节点都需要返回「往左走的最长链长度和往右走的最长链长度的最大值」给父节点，这样父节点才知道往这边走的最长链的长度是多少。

# 前置知识：树的直径
# 树的直径可以用两次 DFS 或者树形 DP，我在 周赛 328 的第四题讲了树形 DP 的做法，在二叉树直径的基础上略作修改。

# 本题结合了 78 题和 1245 题：枚举城市的子集（子树），求这棵子树的直径。
# 需要注意的是，枚举的子集不一定是一棵树，可能是森林（多棵树，多个连通块）。
# 我们可以在计算树形 DP 的同时去统计访问过的点，看看是否与子集相等，只有相等才是一棵树。

class Solution:
    def countSubgraphsForEachDiameter(self, n: int, edges: List[List[int]]) -> List[int]:
        # 建树
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x - 1].append(y - 1)
            g[y - 1].append(x - 1)  # 编号改为从 0 开始

        ans = [0] * (n - 1)
        in_set = [False] * n
        def f(i: int) -> None:
            if i == n:
                vis = [False] * n
                diameter = 0
                for v, b in enumerate(in_set):
                    if not b: continue
                    # 求树的直径
                    def dfs(x: int) -> int:
                        nonlocal diameter
                        vis[x] = True
                        max_len = 0
                        for y in g[x]:
                            if not vis[y] and in_set[y]:
                                ml = dfs(y) + 1
                                diameter = max(diameter, max_len + ml)
                                max_len = max(max_len, ml)
                        return max_len
                    dfs(v)
                    break
                if diameter and vis == in_set:
                    ans[diameter - 1] += 1
                return
            
            # 不选城市 i
            f(i + 1)

            # 选城市  i
            in_set[i] = True
            f(i + 1)
            in_set[i] = False  # 恢复现场
            
        f(0)
        return ans
    
    
class Solution:
    def countSubgraphsForEachDiameter(self, n: int, edges: List[List[int]]) -> List[int]:
        # 建树
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x - 1].append(y - 1)
            g[y - 1].append(x - 1)  # 编号改为从 0 开始

        ans = [0] * (n - 1)
        
        #  二进制枚举
        for mask in range(3, 1 << n):
            if (mask & (mask - 1)) == 0:  # 需要至少两个点
                continue
            # 求树的直径
            vis = diameter = 0
            
            def dfs(x: int) -> int:
                nonlocal vis, diameter
                vis |= 1 << x  # 标记 x 访问过
                max_len = 0
                for y in g[x]:
                    if (vis >> y & 1) == 0 and mask >> y & 1:  # y 没有访问过且在 mask 中
                        ml = dfs(y) + 1
                        diameter = max(diameter, max_len + ml)
                        max_len = max(max_len, ml)
                return max_len
            dfs(mask.bit_length() - 1)  # 从一个在 mask 中的点开始递归
            if vis == mask:
                ans[diameter - 1] += 1
        return ans


class Solution:
    def countSubgraphsForEachDiameter(self, n: int, edges: List[List[int]]) -> List[int]:
        # 建树
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x - 1].append(y - 1)
            g[y - 1].append(x - 1)  # 编号改为从 0 开始

        # 计算树上任意两点的距离
        dis = [[0] * n for _ in range(n)]
        def dfs(x: int, fa: int) -> None:
            for y in g[x]:
                if y != fa:
                    dis[i][y] = dis[i][x] + 1  # 自顶向下
                    dfs(y, x)
        for i in range(n):
            dfs(i, -1)  # 计算 i 到其余点的距离

        def dfs2(x: int, fa: int) -> int:
            # 能递归到这，说明 x 可以选
            cnt = 1  # 选 x
            for y in g[x]:
                if y != fa and \
                   (di[y] < d or di[y] == d and y > j) and \
                   (dj[y] < d or dj[y] == d and y > i):  # 满足这些条件就可以选
                    cnt *= dfs2(y, x)  # 每棵子树互相独立，采用乘法原理
            if di[x] + dj[x] > d:  # x 是可选点
                cnt += 1  # 不选 x
            return cnt
        
        ans = [0] * (n - 1)
        for i, di in enumerate(dis):
            for j in range(i + 1, n):
                dj = dis[j]
                d = di[j]
                ans[d - 1] += dfs2(i, -1)
        return ans

File: src/1617_Count_Subtrees_With_Max_Distance_Between_Cities.rs
// 1617_Count_Subtrees_With_Max_Distance_Between_Cities
// https://leetcode.cn/problems/count-subtrees-with-max-distance-between-cities/

// 3/12/2023 每日一题
/// 树上任意两节点之间最长的简单路径即为树的「直径」，可以参考「树的直径」、「二叉树的直径」等相关解法。
/// 一颗树可以有多条直径，但直径的长度都是一样的，计算树的直径长度有常见的两种方法:
/// - DP (最远距离 + 次远距离 )
/// - 深度优先搜索 (两次DFS x -> y -> z) 节点 y 与 节点 z 之间的距离即为直径的长度。

use std::collections::VecDeque;
impl Solution {
    pub fn count_subgraphs_for_each_diameter(n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n = n as usize;
        
        // Build the tree
        let mut g = vec![VecDeque::new(); n];
        for edge in &edges {
            let x = (edge[0] - 1) as usize;
            let y = (edge[1] - 1) as usize;
            g[x].push_back(y);
            g[y].push_back(x);
        }

        let mut ans = vec![0; n - 1];
        
        // Binary enumeration
        for mask in 3..(1 << n) {   // 需要至少两个点
            if mask & (mask - 1) == 0 {
                continue;
            }
            
            // 求树的直径
            let mut vis = 0;
            let mut diameter = 0;

            fn dfs(x: usize, g: &Vec<VecDeque<usize>>, mask: usize, vis: &mut usize, diameter: &mut usize) -> i32 {
                *vis |= 1 << x; // 标记 x 访问过

                let mut max_len = 0;
                for &y in &g[x] {
                    if (*vis >> y & 1) == 0 && (mask >> y & 1) == 1 { // y 没有访问过且在 mask 中
                        let ml = dfs(y, g, mask, vis, diameter) + 1;
                        *diameter = (*diameter).max(max_len as usize + ml as usize);
                        max_len = max_len.max(ml);
                    }
                }
                max_len
            }

            // 从一个在 mask 中的点开始递归
            dfs((mask as usize).trailing_zeros() as usize, &g, mask, &mut vis, &mut diameter); 

            if vis == mask {
                ans[diameter - 1] += 1;
            }
        }

        ans
    }
}


impl Solution {
    /// 暴力枚举 i 和 j 作为直径的两个端点 ，那么从 i 到 j 的这条简单路径是直径，这上面的每个点都必须选。
    /// 这条路上还有哪些点是可以选的，哪些点是不能选的？
    pub fn count_subgraphs_for_each_diameter(n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n = n as usize;
        let mut g = vec![vec![]; n];
        for edge in edges {
            g[edge[0] as usize - 1].push(edge[1] as usize - 1);
            g[edge[1] as usize - 1].push(edge[0] as usize - 1);
        }
        let mut result = vec![0; n - 1];
        
        let mut dis = vec![vec![0; n]; n];
        fn dfs(g: &Vec<Vec<usize>>, start: usize, u: usize, fa: usize, dis: &mut Vec<Vec<i32>>) {
            for &v in &g[u] {
                if v != fa {
                    dis[start][v] = dis[start][u] + 1;
                    dfs(g, start, v, u, dis);
                }
            }
        }
        for i in 0..n {
            dfs(&g, i, i, n, &mut dis);
        }
        fn dfs2(g: &Vec<Vec<usize>>, u: usize, fa: usize, dis: &Vec<Vec<i32>>, i: usize, j: usize, d: i32) -> i32 {
            let mut cnt = 1;  // 选u
            for &v in &g[u] {
                if v != fa {
                    if (dis[i][v] < d || dis[i][v] == d && v > j) && (dis[j][v] < d || dis[j][v] == d && v > i) {
                        cnt *= dfs2(g, v, u, dis, i, j, d);  // 这个子树的可选数，乘法原理
                    }
                }
            }
            if dis[i][u] + dis[j][u] > d {
                cnt += 1; // 不选u
            }
            cnt
        }
        for i in 0..n {
            for j in i + 1..n {
                let d = dis[i][j];
                result[d as usize - 1] += dfs2(&g, i, n, &dis, i, j, d);
            }
        }
        result
    }
}